### ISSUE TITLE
Performance & display regressions in matplotlib plotting

### ISSUE THREAD

**Issue by boothby** [networkx/networkx#4806]
I was using the current networkx repo at work yesterday, and did a little bit of plotting.  Below is some code that highlights two significant issues.  The first is that `draw_networkx_edges` has incurred a significant performance regression -- what takes less than a second in networkx 2.5 has exploded to almost a minute now.  The second is that the edges don't touch the nodes.  This plot is admittedly noisy, but the current version is completely illegible.

### Steps to Reproduce
Requires the `dwave_networkx` package to generate the graph I'm plotting

```python
import dwave_networkx as dnx, networkx as nx, time, random
from matplotlib import pyplot as plt
def draw_pegasus(p):
    pos = dnx.drawing.pegasus_layout(p, crosses=True)
    nx.draw_networkx_edges(p, pos, arrows=False)
    nx.draw_networkx_nodes(p, pos, node_size=20, linewidths=0)

p = dnx.pegasus_graph(16) 
plt.figure(figsize=(20, 20)) 
draw_pegasus(p) 
plt.savefig('yield.png') 
plt.close() 
```

### Environment

Python version: 3.8
NetworkX version: 2.5 vs current main branch

### Additional context
I suspect that this was introduced in #4360; and that a solution to the performance regression may be to use `LineCollection` when `arrows=False`.

Networkx 2.5
![25](https://user-images.githubusercontent.com/569654/118373888-f320ae00-b56d-11eb-9cd5-da0205ef60c2.png)

networkx/main
![26pre](https://user-images.githubusercontent.com/569654/118373894-f6b43500-b56d-11eb-9aa0-caaaeb804dd4.png)

**Comment by rossbar** [comment]
> I suspect that this was introduced in #4360; and that a solution to the performance regression may be to use LineCollection when arrows=False.

Yup, that's exactly right. The rationale behind the switch was that `draw_networkx_edges` would always return `FancyArrowPatch`es now instead of different objects depending on whether the graph is undirected or directed. This bit us in `nx-guides` as well: networkx/nx-guides#23.

I wonder if we shouldn't revisit that decision: the performance regression is so severe that it makes `nx_pylab` much less practical for "quick" visualizations.

**Comment by dschult** [comment]
I'm good with going back to different return types depending on whether arrows are requested or not.

BUT...  there are two issues that switching to FancyArrows fixed for us that it would be good to 
revisit before reverting.  

#3337 points out that LineCollections can be iterated over. 
#3398 points out that LineCollections don't allow curved edges.

Those are both nice features of FancyArrows.  Should we make two draw_edges functions -- or an argument to select?  That way a return type and capabilities aren't restricted to a certain graph type.  Default choices could be of course.

**Comment by rossbar** [comment]
Just to put a data point on the severity of the performance regression, here's a simple benchmark for undirected graphs with 1225 edges:

**Before:**

```
In [1]: import networkx as nx

In [2]: G = nx.complete_graph(50)

In [3]: %timeit nx.draw_networkx_edges(G, pos=nx.circular_layout(G))
8.08 ms ± 1.69 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**After:**

```
In [1]: import networkx as nx

In [2]: G = nx.complete_graph(50)

In [3]: %timeit nx.draw_networkx_edges(G, pos=nx.circular_layout(G))
1.31 s ± 7.65 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

That's O(100x) slower for a use-case that is probably quite common (drawing edges for graphs with ~1000 or more edges).

**Comment by rossbar** [comment]
I attempted to undo the performance regression but *keep* the self-loop drawing added in #4370 via reverting, but doing so gets quite complicated very quickly, and I haven't been successful.

The performance regression is really bad, but it would be a shame to lose the self-loop drawing. I'm instead going to pursue an alternative that adds a kwarg to `draw_networkx_edges` to allow the user to toggle between using LineCollections or FancyArrowPatches to try to keep the best of both worlds. Other ideas welcome!

**Comment by dschult** [comment]
We've got a lot of keywords for edge style already:  `style`, `arrowstyle`, `arrowsize`, `arrows`, `connectionstyle`.  Do we already have a way to control the output using these keywords? Or is the logic tree too complex without a single keyword for Line vs Arrow?

Looking at the code, I might have answered this (at least from my perspective):
Can we co-opt the `arrows` keyword?  I don't think this loses any functionality.

Here's the current behavior:
```
arrows : bool, optional (default=True)        
    For directed graphs, if True, set default to drawing arrowheads.        
    Otherwise set default to no arrowheads. Ignored if `arrowstyle` is set.
```
If we are going to allow FancyArrowPatch for undirected networks then one approach is:
- remove the "For directed graphs" part
- remove the "Ignored if arrowstyle is set" logic

So something like: 
```
arrows : bool, optional (default=True)        
    If True, use FancyArrowPatch for the edges. Note that, if desired, `arrowstyle` can turn
    off arrowheads while still allowing the other features of FancyArrowPatch.
    If False, use LineCollection for the edges, which does not allow arrows or bending edges
    including self-loops, but is much faster when you have a lot of edges.
```
I suppose we could change the default to something fancy like:  `(default= (G.size() < 1000))`

**Comment by rossbar** [comment]
One complication is graphs with self-loops, which have to be drawn with FancyArrowPatches - I hadn't considered re-purposing the existing kwargs, but I think you're right that it's worth a look.

**Comment by dschult** [comment]
Right!  If the user wants to show self-loops they need to choose FancyArrowPatch.

We could add a line stating that `arrows` must be True to show self-loops.  
I snuck that info in my example, but not in a very obvious way. :}
```
arrows : bool, optional (default=True)        
    If True, use FancyArrowPatch for the edges. 
        Note that: if desired, `arrowstyle` can turn off arrowheads while still allowing
                          the other features of FancyArrowPatch.
    If False, use LineCollection for the edges.
        This does not allow arrows or curved edges but is much faster for lots of edges.
    Note: you must use FancyArrowPatch to be able to see self-loops.
```
There may be another better way...  but It would be good if we were able to use the existing keywords.

**Comment by rossbar** [comment]
Should be fixed by #4825
